### PR TITLE
Predictable random number seeding across platforms

### PR DESCRIPTION
**Pull request by halflearned**
Another attempting at solving issue #379. 

**Major changes**
+ [C++] Added [`Distributions.h`](https://github.com/halflearned/grf/blob/random/core/src/sampling/Distributions.h) containing uniform and poisson distributions. These were adapted from boost.
+ [C++] All dependencies of the `<random>` header were removed, as well as calls to `std::shuffle`. The most affected file is [`RandomSampler.cpp`](https://github.com/halflearned/grf/blob/random/core/src/sampling/RandomSampler.cpp)
+ ~[C++] Some modification for seeding within each thread in [`ForestTrainer.cpp`](https://github.com/halflearned/grf/blob/master/core/src/forest/ForestTrainer.cpp). This is still not entirely successful, so ensuring more similar results still requires setting the number of threads to 1.~ (This proved unnecessary)

**Note:** The tests will likely _not_ pass at this point. I'm not sure what to do about it, will contact @jtibshirani over slack.

### CODE REVIEW

File: core/src/sampling/RandomSampler.cpp [hunk 1]
```diff
@@ -66,9 +67,9 @@ void RandomSampler::subsample(const std::vector<size_t>& samples,
                               std::vector<size_t>& subsamples,
                               std::vector<size_t>& oob_samples) {
   std::vector<size_t> shuffled_sample(samples);
-  std::shuffle(shuffled_sample.begin(), shuffled_sample.end(), random_number_generator);
+  shuffle(shuffled_sample);
 
-  size_t subsample_size = (size_t) std::ceil(samples.size() * sample_fraction);
+  auto subsample_size = (size_t) std::ceil(samples.size() * sample_fraction);
```
[the remark this answers is not in the document]
Comment (halflearned) [reply core/src/sampling/RandomSampler.cpp]: Sure. I had changed it after observing of a clang-tidy complaint, I can put it back the way it was for consistency with the rest of the codebase.

Screen Shot 2019-07-14 at 11 00 50 AM

File: core/src/forest/ForestOptions.cpp [hunk 2]
```diff
@@ -15,9 +15,9 @@
   along with grf. If not, see <http://www.gnu.org/licenses/>.
  #-------------------------------------------------------------------------------*/
 #include <thread>
-#include <random>
 #include "forest/ForestOptions.h"
 #include "tree/TreeOptions.h"
+#include <random>
```
[the remark this answers is not in the document]
Comment (halflearned) [reply core/src/forest/ForestOptions.cpp]: Good catch, thanks!

File: core/src/sampling/Distributions.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ #------------------------------------------------------------------------  */
+
+#ifndef DISTRIBUTIONS_H
+#define DISTRIBUTIONS_H
+#include <assert.h>
+#include <random>
+#include <vector>
+#include <array>
+
+namespace distributions {
```
[the remark this answers is not in the document]
Comment (halflearned) [reply core/src/sampling/Distributions.h]: Moved it to `third_party`. Still need to add note.

File: r-package/grf/tests/testthat.R [hunk 4]
```diff
@@ -1,4 +1,5 @@
 library(testthat)
 library(grf)
+set.seed(12345)
```
[the remark this answers is not in the document]
Comment (halflearned) [reply r-package/grf/tests/testthat.R]: Reverted all tests.

**Comment by halflearned** [comment]
Thanks for the review, @jtibshirani!

I stripped down the PR to its bare minimum changes, including reverting the unnecessary test seeding. As far as I've been able to test, this set of changes should be able to give us comparable predictions across `clang` and `gcc`. Here's an example:

clang_vs_gcc

The remaining discrepancies are likely due to how both compilers are treating floating point precision. If we round the dependent variable to 12 decimal places, the point predictions are the same, and the estimated treatment effect is also equal up to 15 digits.

clang_vs_gcc_rounded

As before, some tests still fail, but I think this is not this PR's fault. What's the recommendation regarding these tests?

File: core/third_party/Distributions/README.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Distributions
+
+These functions were adapted from `Boost`, version 1.70.0 (released April 12th, 2019 06:04 GMT).
```
Comment (jtibshirani) [inline core/third_party/Distributions/README.md]: Thanks for adding this README, it's really helpful. Now that I see how the code was incorporated, I'm a little concerned at how different it is from the source in `boost`. With the other third-party libraries, we've been very careful not to make modifications to the files. There are a couple reasons for the concern:
1. The refactors introduce a non-trivial risk of bugs. We don't have any unit tests for these refactored classes, making it possible for a subtle bug to slip in.
2. If we want to bump to a new version of the library (in order to pick up a bug fix or optimization), it is hard to understand how to 're-copy' the new code into our repository.

Would it be possible to copy `boost/random/uniform_real_distribution.hpp` and `boost/random/poisson_distribution.hpp` wholesale, keeping the same directory structure and copying over any needed dependencies as well?

As a last note, we also need to include the boost license as we are redistributing the code/ creating a derivative work.

**Comment by jtibshirani** [comment]
> As before, some tests still fail, but I think this is not this PR's fault. What's the recommendation regarding these tests?

We have two options, I'm fine with either way you prefer to go:
1. Try to address the floating point discrepancies between clang and gcc as part of the PR.
2. Rework the randomized tests to get them to pass, and just document the discrepancy for now in the algorithm reference.

**Comment by jtibshirani** [comment]
Closing in favor of #469.

**Pull request closed without merging**